File: apps/listwebserver/src/models/pcap.js
const mongoose = require('mongoose');
const db = require('../managers/database')('offline');

const PcapSchema = new mongoose.Schema({
    id: {
        type: String,
        unique: true,
        required: true,
    },
    owner_id: {
        type: String,
        unique: false,
        required: true,
    },
    file_name: {
        type: String,
        unique: false,
        required: true,
    },
    capture_file_name: {
        type: String,
        required: true,
    },
    pcap_file_name: {
        type: String,
        required: true,
    },
    capture_date: {
        type: Number,
        required: true,
    },
    date: {
        type: Number,
        required: true,
    },
    analyzed: {
        type: Boolean,
        default: false,
    },
    analyzer_version: {
        type: String,
        required: true,
    },
    error: {
        type: String,
        default: '',
    },
    offset_from_ptp_clock: {
        type: Number,
        default: 0,
    },
    anc_streams: {
        type: Number,
        default: 0,
    },
    audio_streams: {
        type: Number,
        default: 0,
    },
    video_streams: {
        type: Number,
        default: 0,
    },
    ttml_streams: {
        type: Number,
        default: 0,
    },
    total_streams: {
        type: Number,
        default: 0,
    },
    narrow_linear_streams: {
        type: Number,
        default: 0,
    },
    narrow_streams: {
        type: Number,
        default: 0,
    },
    not_compliant_streams: {
        type: Number,
        default: 0,
    },
    wide_streams: {
        type: Number,
        default: 0,
    },
    srt_streams: {
        type: Number,
        default: 0,
    },
    generated_from_network: {
        type: Boolean,
        default: false,
    },
    summary: {
        type: Object,
        default: {},
    },
    analysis_profile: {
        type: mongoose.Schema.Types.Mixed,
    },
    truncated: {
        type: Boolean,
        default: false,
    },
    sdps: {
        type: Array,
        default: []
    },
    parsed_sdps: {
        type: Array,
        default: [],
    },
    media_type_map: {
        type: Array,
        default: [],
    },
    transport_type: {
        type: String,
        default: "RTP"
    }
});

PcapSchema.set('toJSON', {
    versionKey: false,
});

module.exports = db.model('pcaps', PcapSchema);

File: cpp/libs/analysis/lib/src/ebu/list/analysis/serialization/pcap.cpp
#include "ebu/list/analysis/serialization/pcap.h"

using namespace ebu_list;
using namespace ebu_list::analysis;

nlohmann::json pcap_info::to_json(const pcap_info& info)
{
    nlohmann::json j;
    j["id"]               = info.id;
    j["file_name"]        = info.filename;
    j["pcap_file_name"]   = info.pcap_file_name;
    j["analyzer_version"] = info.analyzer_version;
    j["date"]             = std::chrono::duration_cast<std::chrono::milliseconds>(info.date.time_since_epoch()).count();
    j["transport_type"]   = info.transport_type;
    j["capture_date"] =
        std::chrono::duration_cast<std::chrono::milliseconds>(info.capture_timestamp.time_since_epoch()).count();
    j["analyzed"]  = info.analyzed;
    j["truncated"] = info.truncated;
    j["offset_from_ptp_clock"] =
        std::chrono::duration_cast<std::chrono::nanoseconds>(info.offset_from_ptp_clock).count();
    j["video_streams"]         = info.video_streams;
    j["audio_streams"]         = info.audio_streams;
    j["anc_streams"]           = info.anc_streams;
    j["ttml_streams"]          = info.ttml_streams;
    j["total_streams"]         = info.total_streams;
    j["wide_streams"]          = info.wide_streams;
    j["narrow_streams"]        = info.narrow_streams;
    j["narrow_linear_streams"] = info.narrow_linear_streams;
    j["not_compliant_streams"] = info.not_compliant_streams;
    j["srt_streams"]           = info.srt_streams;

    return j;
}

pcap_info pcap_info::from_json(const nlohmann::json& j)
{
    logger()->info("from json {}", j.dump());
    pcap_info pcap;
    pcap.id             = j.at("id").get<std::string>();
    pcap.filename       = j.at("file_name").get<std::string>();
    pcap.pcap_file_name = j.at("pcap_file_name").get<std::string>();
    pcap.transport_type = j.at("transport_type").get<std::string>();

    auto analyzer_version_json = j.find("analyzer_version");
    if(analyzer_version_json != j.end())
        pcap.analyzer_version = analyzer_version_json->get<std::string>();
    else
        pcap.analyzer_version = "";

    const auto duration        = std::chrono::duration<clock::rep, std::milli>(j.at("date").get<clock::rep>());
    const auto capture_date    = std::chrono::duration<clock::rep, std::milli>(j.at("capture_date").get<clock::rep>());
    pcap.date                  = std::chrono::system_clock::time_point(duration);
    pcap.capture_timestamp     = clock::time_point(capture_date);
    pcap.analyzed              = j.at("analyzed").get<bool>();
    pcap.truncated             = j.at("truncated").get<bool>();
    const auto ptp_offset_ns   = j.at("offset_from_ptp_clock").get<uint64_t>();
    pcap.offset_from_ptp_clock = std::chrono::nanoseconds{ptp_offset_ns};
    pcap.video_streams         = j.at("video_streams").get<int>();
    pcap.audio_streams         = j.at("audio_streams").get<int>();
    pcap.ttml_streams          = j.at("ttml_streams").get<int>();
    pcap.total_streams         = j.at("total_streams").get<int>();
    pcap.wide_streams          = j.at("wide_streams").get<int>();
    pcap.narrow_streams        = j.at("narrow_streams").get<int>();
    pcap.narrow_linear_streams = j.at("narrow_linear_streams").get<int>();
    pcap.not_compliant_streams = j.at("not_compliant_streams").get<int>();
    pcap.srt_streams           = j.at("srt_streams").get<int>();

    return pcap;
}
